### PR TITLE
Add support for anonymous auth

### PR DESCRIPTION
**Pull request by zengyan-amazon**
*Issue #, if available:*

*Description of changes:* Support anonymous auth

see: https://docs.search-guard.com/latest/kibana-authentication-anonymous#using-kibana-with-anonymous-authentication

*Test:
```
% curl http://localhost:5601/jpo/auth/anonymous
{"username":"opendistro_security_anonymous","tenants":{"opendistro_security_anonymous":true},"roles":["own_index"],"backendroles":["opendistro_security_anonymous_backendrole"],"selectedTenants":"__user__"}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/auth/types/basic/routes.ts [hunk 1]
```diff
@@ -165,115 +158,55 @@ export class BasicAuthRoutes {
...
+              roles: user.roles,
+              backendroles: user.backendRoles,
+              selectedTenants: this.config.multitenancy?.enabled
+                ? sessionStorage.tenant
+                : undefined,
             },
           });
+        } else {
+          return response.badRequest({
+            body: 'Anonymouse auth is disabled.',
```
Comment (yuxitang-amzn) [inline server/auth/types/basic/routes.ts:206]: Typo

File: test/jest.config.integration.js [hunk 2]
```diff
@@ -24,7 +24,7 @@ export default {
   ),
   setupFilesAfterEnv: [
     '<rootDir>/src/dev/jest/setup/after_env.integration.js',
-    '<rootDir>/plugins/opendistro_security/test/es/setup_es.js',
+    // '<rootDir>/plugins/opendistro_security/test/es/setup_es.js',
```
Comment (yuxitang-amzn) [inline test/jest.config.integration.js:27]: Remove unused code

File: test/jest_integration/basic_auth.test.ts [hunk 3]
```diff
@@ -24,14 +24,17 @@ import {
...
 } from '../constant';
+import { getAuthCookie, extractAuthCookie } from '../helper/cookie';
 
 describe('start kibana server', () => {
   let root: Root;
   let esProcess: ChildProcess;
 
   beforeAll(async () => {
-    esProcess = await startElasticsearch();
+    // esProcess = await startElasticsearch();
```
Comment (yuxitang-amzn) [inline test/jest_integration/basic_auth.test.ts:37]: Remove unused code

File: test/jest_integration/basic_auth.test.ts [hunk 4]
```diff
@@ -64,7 +67,7 @@ describe('start kibana server', () => {
     // shutdown Kibana server
     root.shutdown();
     // shutdown Elasticsearch
-    await stopElasticsearch(esProcess);
+    // await stopElasticsearch(esProcess);
```
Comment (yuxitang-amzn) [inline test/jest_integration/basic_auth.test.ts:70]: Remove unused code

File: server/backend/opendistro_security_plugin.ts [hunk 5]
```diff
@@ -17,26 +17,49 @@
...
+   * }
+   */
   Client.prototype.opendistro_security.prototype.multitenancyinfo = ca({
     url: {
       fmt: '/_opendistro/_security/kibanainfo',
     },
   });
 
+  /**
+   * Gets tenant info of current user.
```
Comment (yuxitang-amzn) [inline server/backend/opendistro_security_plugin.ts:52]: Could you also add response schema like above?

File: public/plugin.ts [hunk 6]
```diff
@@ -33,7 +33,7 @@ import { setupTopNavButton } from './apps/account/account-app';
 async function hasApiPermission(core: CoreSetup): Promise<boolean | undefined> {
   try {
     const permissions = await core.http.get(API_ENDPOINT_PERMISSIONS_INFO);
-    return permissions?.data?.has_api_access || false;
+    return permissions.has_api_access || false;
```
Comment (yuxitang-amzn) [inline public/plugin.ts:36]: Is this change manually tested?

File: test/jest_integration/basic_auth.test.ts [hunk 7]
```diff
@@ -86,4 +89,51 @@ describe('start kibana server', () => {
...
+    const response = await kbnTestServer.request
+      .post(root, '/auth/login')
+      .unset(AUTHORIZATION_HEADER_NAME)
+      .send({
+        username: ADMIN_USER,
+        password: 'invalid',
+      });
+    expect(response.status).toEqual(401);
+  });
+
```
Comment (yuxitang-amzn) [inline test/jest_integration/basic_auth.test.ts:127]: Is the anonymous login going to be tested?

File: server/auth/types/basic/routes.ts (hunk 1 above)
Comment (zengyan-amazon) [reply server/auth/types/basic/routes.ts:206]: fixed

File: test/jest.config.integration.js (hunk 2 above)
Comment (zengyan-amazon) [reply test/jest.config.integration.js:27]: actually those code are temporary commented out for fast testing, will uncomment them

File: test/jest_integration/basic_auth.test.ts (hunk 3 above)
Comment (zengyan-amazon) [reply test/jest_integration/basic_auth.test.ts:37]: same as above

File: test/jest_integration/basic_auth.test.ts (hunk 4 above)
Comment (zengyan-amazon) [reply test/jest_integration/basic_auth.test.ts:70]: same as above

File: server/backend/opendistro_security_plugin.ts (hunk 5 above)
Comment (zengyan-amazon) [reply server/backend/opendistro_security_plugin.ts:52]: added

File: public/plugin.ts (hunk 6 above)
Comment (zengyan-amazon) [reply public/plugin.ts:36]: yes it is tested

File: test/jest_integration/basic_auth.test.ts (hunk 7 above)
Comment (zengyan-amazon) [reply test/jest_integration/basic_auth.test.ts:127]: sure, will write test case for anonymous auth

File: server/backend/opendistro_security_plugin.ts [hunk 8]
```diff
@@ -17,26 +17,56 @@
...
+  /**
+   * Gets tenant info of current user. The output looks like:
+   * {
+   *   "user_name": "admin",
+   *   "not_fail_on_forbidden_enabled": false,
+   *   "kibana_mt_enabled": true,
+   *   "kibana_index": ".kibana",
+   *   "kibana_server_user": "kibanaserver"
+   * }
+   */
```
Comment (yuxitang-amzn) [inline server/backend/opendistro_security_plugin.ts:60]: These 2 endpoints returns same thing?

Comment (zengyan-amazon) [reply server/backend/opendistro_security_plugin.ts:60]: my bad, updated

File: server/backend/opendistro_security_plugin.ts [hunk 9]
```diff
@@ -17,26 +17,52 @@
...
   Client.prototype.opendistro_security.prototype.multitenancyinfo = ca({
     url: {
       fmt: '/_opendistro/_security/kibanainfo',
     },
   });
 
+  /**
+   * Gets tenant info. The output looks like:
+   * {
+   *   ".kibana_92668751_admin":"__private__"
```
Comment (yuxitang-amzn) [inline server/backend/opendistro_security_plugin.ts:54]: Just to confirm, it is not like `{tenant1: true, tenant2: true}` ?

**Pull request merged**